cmd/complex/logger.go

```diff
@@ -2,77 +2,46 @@ package main
-	l *log.Logger
-}
-
-// Log is a logger for sse.Server. It logs events like so:
-//
-//	level=INFO msg="A message about the logged event"
-//	  ua="the user agent" ip="the RemoteAddr" host="the host" origin="the origin"
-//	  lastEventID=5 topics=""
-//
-// It expects that the given context contains a value mapped to loggerCtxKey{},
-// which will happen because sse.Server passes the request context and we set
-// a middleware up to add these to the context.
-func (l logger) Log(ctx context.Context, level sse.LogLevel, msg string, data map[string]any) {
-	levelStr := ""
-	switch level {
-	case sse.LogLevelInfo:
-		levelStr = "INFO"
-	case sse.LogLevelWarn:
-		levelStr = "WARN"
-	case sse.LogLevelError:
-		levelStr = "ERROR"
-	}
-
-	output := fmt.Sprintf("level=%s msg=%q", levelStr, msg)
-	if e, ok := data["err"]; ok {
-		output += fmt.Sprintf(" err=%q", e.(error).Error())
-	}
-
-	delete(data, "err")
-
-	output += "\n  "
-
-	for k, v := range ctx.Value(loggerCtxKey{}).(map[string]string) {
-		output += fmt.Sprintf(" %s=%q", k, v)
-	}
-
-	output += "\n  "
-
-	for k, v := range data {
-		switch s := v.(type) {
-		case sse.EventID: // key is "lastEventID"
-			output += fmt.Sprintf(" %s=%q", k, s)
-		case []string: // key is "topics"
-			output += fmt.Sprintf(" %s=%q", k, strings.Join(s, ", "))
-		default:
-			panic(fmt.Errorf("unexpected value type %T", v))
-		}
-	}
-
-	l.l.Println(output)
-}
-
 type loggerCtxKey struct{}
 
+// withLogger is an http middleware that generates a logger with request-specific fields
+// added to it and attaches it to the request context for later retrieval with getLogger().
+// This is simmilar to how existing packages like https://github.com/go-chi/httplog works.
 func withLogger(h http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		r = r.WithContext(context.WithValue(r.Context(), loggerCtxKey{}, map[string]string{
-			"ua":     r.UserAgent(),
-			"ip":     r.RemoteAddr,
-			"host":   r.Host,
-			"origin": r.Header.Get("Origin"),
-		}))
+		l := slog.Default().With(
+
+			"UserAgent", r.UserAgent(),
+			"RemoteAddr", r.RemoteAddr,
+			"Host", r.Host,
+			"Origin", r.Header.Get("origin"),
+		)
+		r = r.WithContext(context.WithValue(r.Context(), loggerCtxKey{}, l))
 		h.ServeHTTP(w, r)
 	})
 }
+
+// getLogger retrieves the request-specific logger from a request's context. This is
+// similar to how existing packages like https://github.com/go-chi/httplog work.
+func getLogger(r *http.Request) (*slog.Logger, error) {
+	logger, ok := r.Context().Value(loggerCtxKey{}).(*slog.Logger)
+	if !ok {
+		return nil, errors.New("logger not available on request")
+	}
+
+	return logger, nil
+}
+
+// setupLogger just initializes the global logger in whatever way you might prefer. We are
+// keeping it simple for this example.
+func setupLogger() {
+	h := slog.NewTextHandler(os.Stderr, &slog.HandlerOptions{Level: slog.LevelDebug})
+	logger := slog.New(h)
+	slog.SetDefault(logger)
+}
```

Comment:
Given that `getLogger` is used in just one place let's remove boilerplate and have this function fit for `sse.Server` directly.

Also, do we need a global logger? This code doesn't call the global logger at all. What if we just initialize that logger in `main` and pass it where necessary? We could also have `withLogger` take in the base logger and not rely on `slog.Default()`. No globals should make code easier to follow.

```suggestion
func getLogger(r *http.Request) *slog.Logger {
	return r.Context().Value(loggerCtxKey{}).(*slog.Logger)
}
```

Reply:
In the withLogger function a new logger handle is created from slog.Default(), which returns a pointer to the global logger. Looking at it now, I was planning on adding a log message to the OnSession function which would also use the default logger, since that seems like a likely thing for someone to do in practice, but left it out because it seemed like an unneccesary addition.

I could put all this stuff just in main if you prefer? I just split it up this way to keep the most 1-to-1 translation of what was already there.

Reply:
Oh, I also did add a small thing to demonstrate how to get the http.Server error logs at the same time. That also utilizes the default logger to create a *log.Logger that http.Server.ErrorLog can use which ties the whole logging story together with a single handler.

Reply:
If it's okay to change the structure of the example more significantly, I think what I would tend to do normally would be to make getLogger a closure like `getLogger(*slog.Logger) func(r *http.Request) *slog.Logger`. I would then explicitly pass a *slog.Logger to the newSSE function and create a logger object in main() to pass around.

I don't personally like globals, so that would be my preference anyways.

Reply:
But if `getLogger` returns a closure over the given logger, does it still retrieve a logger from the request context? If it does, then passing a logger through the closure doesn't make sense, as it will not be used.

The `http.Server.ErrorLog` thing is nice, didn't know about it!

Reply:
oh sorry, I meant withLogger could be a closure. That way main() could instantiate a new logger and pass it to withLogger to create a `func(h http.Handler) http.Handler` middleware that creates the per-request loggers from there instead of the global logger.

Reply:
Yes, that's also how I thought of it, let's do it this way.

Reply:
ok, take a look now